### PR TITLE
feat(storage): Implement storage schema RPC de/serializer, merge, APIs

### LINKED ISSUE #13251: Implement storage schema RPC APIs
Dependent on stubs from #13241

## High level tasks

- [x] Extend the `Storage` service definition

The `Storage` service defined in the [protobuf definition](https://github.com/influxdata/influxdb/blob/e9600b1f0be0cced84f376e0a633998e78970bd0/storage/reads/datatypes/storage_common.proto#L15) will need to expose the new `TagKeys` and `TagValues` APIs. These APIs should return a `stream` similar to the [`Read`](https://github.com/influxdata/influxdb/blob/e9600b1f0be0cced84f376e0a633998e78970bd0/storage/reads/datatypes/storage_common.proto#L17) API in order to handle very large responses in an efficient manner. The chunked response would likely be slices of byte slices (`[][]byte`) representing the tag or key values.

----

- [x] Extend the `reads.Store` interface with new `TagKeys` and `TagValues` APIs

The [`reads.Store`](https://github.com/influxdata/influxdb/blob/d0603457b7a2ef876e62364d8667cc7a97e7ca05/storage/reads/store.go#L77-L81) interface is intended to abstract the various physical operations offered by the storage layer. The `TagKeys` and `TagValues` APIs should be added to this interface. The implementation of this interface exists in OSS and Cloud 2.0.

Note that this API returns interfaces also defined in the `reads` package. These abstractions  typically follow an iterator pattern with a `Next` API. The iterator approach allows us to implement in-process or inter-process responses efficiently, as the internal details can batch network requests as appropriate.

See the `reads.ResultSet` details below which walks through the implementation for in-process and inter-process behavior. The process should be similar for the `TagKeys` and `TagValues` APIs, albeit much simpler, given these APIs are ultimately streaming a sequence of string or byte slices. 

The result returned by `TagKeys` or `TagValues` via  `storage.Engine` will be an string iterator similar to:

```go
type StringIterator interface {
    Next() bool    // Returns false it iteration has concluded
    Value() string // Returns the current value.
}
```

A similar definition could exist in the `reads` package that can be reused whenever an RPC needs to return a sequence of strings.

----

- [x] Extend [`readservice.store`](https://github.com/influxdata/influxdb/blob/d0603457b7a2ef876e62364d8667cc7a97e7ca05/storage/readservice/store.go#L17) to implement refactored `reads.Store` interface

This is the OSS implementation of the `reads.Store` interface. A separate issue will be created to track the Cloud 2.0 work.

## `reads.ResultSet`

In order to complete the in-process and inter-process implementation the author will need to consider these four areas, namely: 

* The in-process implementation of the `reads.ResultSet`
* A binary serializer / writer for the in-process implementation, which takes a `read.ResultSet` and write it to a gRPC stream
* A binary deserializer / reader, which implements the `reads.ResultSet` interface and reads from a gRPC stream
* Merging multiple `reads.ResultSet` inputs to handle Cloud and Enterprise

Using `ResultSet` as an example:

* Implementation of the in-process [`ResultSet`](https://github.com/influxdata/influxdb/blob/d73d73c0d4b708462ef58c8851577486aeb3de56/storage/reads/resultset.go#L16)

* [`ResponseWriter`](https://github.com/influxdata/influxdb/blob/d73d73c0d4b708462ef58c8851577486aeb3de56/storage/reads/response_writer.go#L65) is the implementation of the `ResultSet` binary serializer

* The [`WriteResultSet`](https://github.com/influxdata/influxdb/blob/d73d73c0d4b708462ef58c8851577486aeb3de56/storage/reads/response_writer.go#L65) API takes an in-process `ResultSet` and writes it to a gRPC stream, provided to the `ResponseWriter` on initialization

* The [`ResultSetStreamReader`](https://github.com/influxdata/influxdb/blob/da24fe6c7f54697870e062813f91a18a383c7ceb/storage/reads/stream_reader.go#L98) serves as the complimentary binary deserializer of the `ResponseWriter` and implements the `reads.ResultSet` interface.

* The, [`mergedResultSet`](https://github.com/influxdata/influxdb/blob/5f241e5e5e52d272643be079cca2dbe53f3b3096/storage/reads/merge.go#L10-L15) is responsible for merging multiple instances of `ResultSet` into a single stream

### LINKED ISSUE #13271: Implement storage schema RPC serializer / deserializer and merge
Dependent on stubs from #13241

This issue tracks the required implementations of types responsible for serializing, deserializing and merging `StringIterator` instances over gRPC.

See #13251 for additional detail on the definition of `StringIterator` and the `reads.ResultSet` section, which describes the behavior of the following types.

## High level tasks 

- [x] Implement a `StringIterator` gRPC writer / serializer

This type is analagous to the `ResponseWriter` detailed in the `reads.ResultSet` section of #13251, and is reponsible for serializing the `StringIterator` to a gRPC stream

------

- [x] Implement a `StringIterator` gRPC reader / deserializer

This type is analagous to the `ResultSetStreamReader` detailed in the `reads.ResultSet` section of #13251, and is reponsible for deserializing the `StringIterator` from a gRPC stream. This type must implement a `StringIterator` interface or similar.

------

- [x] Implement a `StringIterator` merger

This type is analagous to the `mergedResultSet` detailed in the `reads.ResultSet` section of #13251, and is reponsible for merging and deduplicating multiple `StringIterator` instances into a single `StringIterator`.

### PR DESCRIPTION
**Pull request by jacobmarble**
Closes #13271
Closes #13251
Helps https://github.com/influxdata/idpe/issues/3114

Adds schema support to storage APIs

### CODE REVIEW

File: storage/reads/datatypes/storage_common.proto [hunk 1]
```diff
@@ -24,6 +24,12 @@ service Storage {
 
   rpc Hints (google.protobuf.Empty) returns (HintsResponse);
 
+  // TagKeys performs a read operation for tag keys
+  rpc TagKeys (TagsRequest) returns (stream TagsResponse);
```
Comment (stuartcarnie) [inline storage/reads/datatypes/storage_common.proto]: need distinct `TagKeysRequest` and `TagValuesRequest` arguments to these APIs, as they have different parameters

File: storage/reads/datatypes/storage_common.proto [hunk 2]
```diff
@@ -24,6 +24,12 @@ service Storage {
...
+
+  // TagValues performs a read operation for tag values
+  rpc TagValues (TagsRequest) returns (stream TagsResponse);
```
Comment (stuartcarnie) [inline storage/reads/datatypes/storage_common.proto]: Add `TagValuesRequest` that has the same parameters as `TagKeysRequest` with the addition of `string tag_key` to specify which tag key the values should be produced for

File: storage/reads/datatypes/storage_common.proto [hunk 3]
```diff
@@ -208,6 +214,18 @@ message TimestampRange {
...
 
+// TagsRequest is the request message for the Storage.TagKeys and Storage.TagValues
+message TagsRequest {
+  google.protobuf.Any tags_source = 1 [(gogoproto.customname) = "TagsSource"];
+  TimestampRange range = 2 [(gogoproto.nullable) = false];
+  Predicate predicate = 3;
+}
+
+// Response message for Storage.TagKeys and Storage.TagValues.
+message TagsResponse {
```
Comment (stuartcarnie) [inline storage/reads/datatypes/storage_common.proto]: This should be renamed to something more generic, like `StringValuesResponse` given it represents the generic `StringIterator` and its associated merge operation.

File: storage/reads/merge.go [hunk 4]
```diff
@@ -114,10 +115,47 @@ func (h *resultSetHeap) Push(x interface{}) {
...
+	}
+}
+
+func (mr *MergedStringIterator) Next() bool {
+	for len(mr.iterators) > 0 {
+		iterator := mr.iterators[0]
+
+		for iterator.Next() {
+			mr.nextValue = iterator.Value()
+			if _, found := mr.uniqueValues[mr.nextValue]; !found {
```
Comment (stuartcarnie) [inline storage/reads/merge.go]: A `StringIterator` returned by the engine's `TagValues` or `TagKeys` APIs will produce sorted results, so you can rely on that to implement a k-way merge sort. 

Given there is still a lot to do, we can add that as a TODO in this code and come back to it later this week with a separate PR.

Comment (jacobmarble) [reply storage/reads/merge.go]: I tried to find that written down somewhere, because I thought you mentioned it verbally. Anyway, #13440

**Pull request merged**